### DISCUSSION TITLE
Alternate to org.eclipse.glsp.server.features.directediting.ApplyLabelEditOperation

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#649] [Q&A] [answered]
I am using GLSP's vscode intergration project.

Also, I am using Xtext with a language sever and language client(Generating diagrams using GLSP). Data required to generate graph will be provided by language server.

I want to implement label editing. In case of GLSP when we try to edit diagram element's label, org.eclipse.glsp.server.features.directediting.ApplyLabelEditOperation will be sent to server from the client where we will update GLabel's text.

**My requirement:** Language resource's element corresponding to edited diagram element need to be updated.
To update Language resource's element we have to invoke rename from vscode(client) side. But org.eclipse.glsp.server.features.directediting.ApplyLabelEditOperation will reach server side. Any way to avoid this so that we can handle ApplyLabelEditOperation at client side? or Can we invoke any other Action/Command instead of **ApplyLabelEditOperation** when label is edited?

I want to know if there are any other ways to implement my requirement.

**Comment by planger** [comment, marked as answer]
I can think of three ways to address this requirement:

1. In use cases like you have, we would typically have a connection from the GLSP server to your language server, so that you can query its internal state, like the AST, using custom operations and then create a diagram from it. If you have the language server connection, you could invoke the rename method in the language server also from the GLSP server.

2. Register an additional action handler on the client that reacts to `ApplyLabelEditOperation` and then call the language server from there. On the GLSP server, just register a noop for the label editing handler. This would look something like the following in the client code:

```typescript
export const yourModule = new ContainerModule((bind, _unbind, isBound) => {
    configureActionHandler({ bind, isBound }, ApplyLabelEditOperation.KIND, ApplyLabelEditOperationHandler);
});

@injectable()
export class ApplyLabelEditOperationHandler implements IActionHandler {
    handle(action: Action): void | Action | ICommand {
        console.log('Do what you have to do');
    }
}

...
export default function createContainer(widgetId: string): Container {
   const container = createClientContainer(yourDiagramModule, yourApplyLabelEditModule);
   ...
}

```

3. Customize the module that introduces the label editing and adjust which action is triggered on [apply of a label edit](https://github.com/eclipse/sprotty/blob/6f60f2de3a6e51e63188528076b88c87f6f23bea/packages/sprotty/src/features/edit/edit-label-ui.ts#L123).

**Comment by sathya-1994** [comment]
@planger ,

Thank you, I am able to implement diagram editing using approach 1.
